[WIP] Louvain algorithm for undirected graphs

jpacold opened the pull request:
I finally had some time to work on #1141, so far only on the Rust side and only for undirected graphs. Opening this for feedback on how the Rust code is organized, and maybe to restart the discussion on whether this should go here or in petgraph.

The big additions here are: 

* `metrics.rs`
    * `Modularity`: a trait for graphs for which we can compute [modularity](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.community.quality.modularity.html#networkx.algorithms.community.quality.modularity) (by extension this means we can apply the Louvain method)
    * `modularity`: computes the modularity, given a graph and a set of subsets.
    * `Partition`: a struct holding a graph partition. This mainly exists to keep the implementation of `modularity` organized.

* `louvain.rs`
    * `InnerGraph`: at each level of the algorithm we work with an aggregated graph, where each node of this graph corresponds to one of the communities that were identified in the previous level). `InnerGraph` keeps track of this correspondence.
    * `LouvainAlgo`: helper functions for updating the `Partition` at each level of the algorithm.
    * `one_level_undirected`: most of the business logic is in here since it constructs the new communities at each level, roughly equivalent to `_one_level` in the [networkx implementation](https://networkx.org/documentation/stable/_modules/networkx/algorithms/community/louvain.html).
    * `louvain_communities`: repeatedly calls `one_level_undirected`, stopping either after a fixed number of iterations or if the modularity improvement falls below the given threshold.

I realize this is a lot to review at once and can go into more detail about each piece if necessary.

IvanIsCoding commented on rustworkx-core/src/community/louvain.rs:
```diff
@@ -0,0 +1,432 @@
+            } else {
+                0.0
+            };
+            let remove_cost = -delta / m + resolution * (s_tot[best_com] * deg) / two_m_sq;
```
You might want to make this expression a closure. Just to keep it consistent inside and outside the loop

jpacold replied:
Done in 12ffc33.

IvanIsCoding commented on rustworkx-core/src/community/louvain.rs:
```diff
@@ -0,0 +1,432 @@
+            }
+        }
+        if !performed_move {
+            break;
```
I think we should loop on `while performed_move` or something like that to make the intention clearer.

jpacold replied:
Done in 234659e.

IvanIsCoding commented on rustworkx-core/src/community/louvain.rs:
```diff
@@ -0,0 +1,432 @@
+    }
+
+    #[test]
+    fn test_louvain_karate_club_graph() {
```
Friendly reminder to myself that we should add a Karate club graph generator. It could be a separate PR

IvanIsCoding replied:
Once #1280 gets merged this test will hopefully be much shorter

IvanIsCoding commented on rustworkx-core/src/community/louvain.rs:
```diff
@@ -0,0 +1,432 @@
+        // happen to get the same result as:
+        //      import networkx as nx
+        //      g = nx.karate_club_graph()
+        //      communities = nx.community.louvain_communities(g, weight='weight', seed=12)
```
I think this is a brittle test (https://testing.googleblog.com/2024/04/how-i-learned-to-stop-writing-brittle.html) but leave a TODO there and we can revisit it later. I am glad it passes though

jpacold replied:
I'm not sure yet what to do with this. I can adjust the resolution parameter so that we consistently get two output communities. However, there is still a lot of dependence on the seed (in networkx I get at least 5 different partitions with the same parameters)

IvanIsCoding commented on rustworkx-core/src/community/louvain.rs:
```diff
@@ -0,0 +1,432 @@
+    for n in input_graph.node_identifiers() {
+        let prev_com = partition.subset_idx(n);
+        let inner_com = node_to_community[prev_com];
+        let new_com = if let Some(&c) = final_index.get(&inner_com) {
```
Nitpicking but a `match` statement would be clearer in this case

jpacold replied:
Done in 11a74fd.

IvanIsCoding commented on rustworkx-core/src/community/metrics.rs:
```diff
@@ -0,0 +1,233 @@
+        };
+
+        let m: f64 = total_edge_weight(self.graph);
+        sigma_internal / m - resolution * sigma_total_squared / (m * m)
```
We might want to reorganize this expression to avoid floating point erros, specially because of the common factor `1/m`

jpacold replied:
I started doing this in 58857fd. Other than the common factor what are the possible issues? I think `sigma_internal`, `sigma_total_squared / m`, and `m` are all about the same order of magnitude but could be missing something.

IvanIsCoding reviewed (commented):
Overall this is a great start! I would call it more than a draft given it already has good tests.

I left some Rust comments, I need to study the algorithm more myself to give more "algorithmic" advice

jpacold commented:
* Fixed two mistakes in the modularity gain calculation (I am actually a little surprised that the tests passed earlier and need to come up with some more)
* Addressed some of the comments

It looks as if the CI is failing because I used [associated type bounds](https://blog.rust-lang.org/2024/06/13/Rust-1.79.0.html#bounds-in-associated-type-position).

IvanIsCoding commented:
> * Fixed two mistakes in the modularity gain calculation (I am actually a little surprised that the tests passed earlier and need to come up with some more)
> * Addressed some of the comments
> 
> It looks as if the CI is failing because I used [associated type bounds](https://blog.rust-lang.org/2024/06/13/Rust-1.79.0.html#bounds-in-associated-type-position).

Indeed our current MSRV is Rust 1.70 from June 2023. Rust 1.79 is from June 2024. In the past we used to follow Debian stable's Rustc which believe it or not is in [1.63 still](https://packages.debian.org/trixie/rustc).

I'd try rewriting it before we discuss bumping MSR, 1.79 might break a couple people that install from source
